// bitfinex-api-node/examples/rest2/symbols.js
'use strict'

process.env.DEBUG = 'bfx:examples:*'

const debug = require('debug')('bfx:examples:rest2_symbols')
const bfx = require('../bfx')
const rest = bfx.rest(2)

var symbol_array = [],
    ETH_Pairs = [],
    BTC_Pairs = [],
    USD_Pairs = [],
    EUR_Pairs = [],
    GBP_Pairs = [],
    JPY_Pairs = [],
    ETHBTC_pairs = [],
    USDBTC_pairs = [],
    USDETH_pairs = [],
    EURBTC_pairs = [],
    EURETH_pairs = [],
    GBPBTC_pairs = [],
    GBPETH_pairs = [],
    JPYBTC_pairs = [],
    JPYETH_pairs = [];

var e = 'ETH', b = 'BTC', u = 'USD', r = 'EUR', g = 'GBP', j = 'JPY';
var key, value, result;

var makeArray = function (sub){
  var array_s = symbol_array[0]
  array_s.forEach(val => {
    if(val.substring(4) == sub){
      
      if(sub == e) {ETH_Pairs.push(val)}
      if(sub == b) {BTC_Pairs.push(val)}
      if(sub == u) {USD_Pairs.push(val)}
      if(sub == r) {EUR_Pairs.push(val)}
      if(sub == g) {GBP_Pairs.push(val)}
      if(sub == j) {JPY_Pairs.push(val)}

      debug('Added',val, 'to', sub, 'array')
    }
  })
  debug('Done', sub, '\n')
}

var compareArray = function (a1, a2, finalarr){
  a1.forEach(val =>{
    a2.forEach(val2 =>{ 
      var alt = val2.substring(0,4)
      if(val.includes(alt)){
        finalarr.push(val)
        finalarr.push(val2)
        debug('Pushed %s and %s', val, val2)
      }
    })
  })
  finalarr.push('tETHBTC')
}

debug('fetching symbol list...')

rest.symbols().then(symbols => {
  symbol_array.push(symbols.map(s => `t${s.toUpperCase()}`))
  //debug(symbol_array)
  debug('available symbols are: %s', symbol_array)
  makeArray(e);
  makeArray(b);
  makeArray(u);
  makeArray(g);
  makeArray(j);
  compareArray(ETH_Pairs, BTC_Pairs, ETHBTC_pairs);
  compareArray(USD_Pairs, BTC_Pairs, USDBTC_pairs);
  compareArray(USD_Pairs, ETH_Pairs, USDETH_pairs);
  compareArray(GBP_Pairs, BTC_Pairs, GBPBTC_pairs);
  compareArray(GBP_Pairs, ETH_Pairs, GBPETH_pairs);
  compareArray(EUR_Pairs, BTC_Pairs, EURBTC_pairs);
  compareArray(EUR_Pairs, ETH_Pairs, EURETH_pairs);
  //debug(ETHBTC_pairs)
}).catch(err => {
  debug('error: %s', err)
})

module.exports.symbol_array = symbol_array;
module.exports.ethpairs = ETH_Pairs;
module.exports.btcpairs = BTC_Pairs;
module.exports.usdpairs = USD_Pairs;
module.exports.eurpairs = EUR_Pairs;

module.exports.ethbtcpairs = ETHBTC_pairs;
module.exports.usdbtcpairs = USDBTC_pairs;
module.exports.usdethpairs = USDETH_pairs;
module.exports.eurbtcpairs = EURBTC_pairs;  
module.exports.eurethpairs = EURETH_pairs;
